lib/validation/rules/hms.ts

```diff
@@ -11,7 +11,6 @@ export const message = {
 export function hms(required?: HmsType[], msg?: string) {
   return createRule({
     message: ({ lang }) => msg ?? message[lang],
-    optional: true,
     validation: (value) => baseHms(value, required)
   })
 }
```

Comment:
FYI: Vuelidate's `helpers.req()` is internally used in its `required` rule.
https://github.com/vuelidate/vuelidate/blob/08aa51495a4590c17d407a56cbbada9cf370c5d7/packages/validators/src/raw/required.js

Reply:
Ah, wait, I'm a little bit being confused.
`{}` is invalid, so returning `false` is fine, isn't it?

Reply:
First message might be confusing.
Let me summarize it again.

- Vuelidate's helpers.req() method returns `false` for an empty plain object `{}` ([ref](https://github.com/vuelidate/vuelidate/blob/08aa51495a4590c17d407a56cbbada9cf370c5d7/packages/validators/src/raw/core.js#L21-L24))
- If `optional: true` is set, `createRule` returns `true` when `!helpers.req(value)` ([ref](https://github.com/globalbrain/sefirot/blob/745c61b66d58a901d66fbb923a26229ad1e8f35a/lib/validation/Rule.ts#L25))
- So, with `optional: true`, hms rule return `true` for `{}`, which is not expected
- Sefirot's required() validator behaves differently than Vuelidate's helpers.req()
  - It returns `true` for `{}`
  - But, other behavior (e.g. for string) is also different and we come to modify other rules depending on the current `optional: true` behavior
  - This is why I removed `optional: true` from hms rule and add special guard for `undefined` and `null` at validator function level

The question is, is my approach reasonable or should we use Sefirot's `required` function for `optional: true` and modify other validator functions than hms and ymd to be compatible with it?

Reply:
What I'm concerned with is that, if it's fine to add special guard for undefined and null at validator function level, `optional` flag might come to be kinda inconsistent and somehow confusing to use.

Reply:
For this particular validation rule, `hms`, we don't have worry so much about how required works because this validation is only meant to use with `Hms` type which always should include h, m, and s key.

So, most of your concern wouldn't happen on real world anyway 👀  So, let's just `optional: true` here for consistency (wouldn't matter really).

For the `required` concern, yeah we should stick to 1 rule. So, probably we should make `required` rule exactly the same as `helpers.req` and use `required` everywhere I guess 🤔

Reply:
@kiaking 

Thank you for feedback.

So, do you mean to add `optional: true` for hms and ymd rules at the moment with special guard kept at validator function level, and will replace `helpers.req` with our `required` function and fix affected rules as needed in future?

Reply:
Mostly yes, but for validator, I don't think this is "special" guard. At validator level, the use of `required` doesn't have to mandatory. Because whether the value, which is `unknown` should pass or not depends on each validator function.

For example, for `ymd`, I think it should return `false` rather than `true`. Because it's clearly not the expected type. So I think we should remove the comments because that's doing nothing wrong there. Every validator should check if the `value` is expected type or not (regardless of using `required` or how it behaves).

And yes, we should replace `helpers.req` with `required` 👀

Reply:
I replaced `helpers.req()` with internal code: 035e0f3

Note that, I modified `required()` validator and newly defined  `_required()` internal validator.

This is because, space-only string (e.g. `" "`) is treated as invalid by `required` validator because it trims the string, and email rule which has `optional: true` return `true` for such string.

I made `required()` validator trim a string and then call `_required()` with it, so that `optional: true` works as expected. Actually, this is how Vuelidate behaves.